### PR TITLE
fix(library): correct type inference in `ArgsAction`

### PR DESCRIPTION
**Pull request by shulaoda**
Actually, I found that `type Output = v.InferOutput<typeof ComplexSchema>` will be deduced as:
```ts
type Output = {
    fn: (args_0: string) => boolean;
    argsWithFn: (args_0: (...args: unknown[]) => unknown) => unknown;
}
```
Is this the expected behavior?  Is there a problem with my writing for the `args_0` of the `argsWithFn`?

_Originally posted by @shulaoda in https://github.com/fabian-hiller/valibot/issues/968#issuecomment-2537846099_

The second generic of `BaseTransformation` should be `InferOutput<...>`.

Playground: https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6BjCAO1XgGF8xSBTAB4BlJgAs+IZHAC8mAHQcowJjADyFAFZ8VACgDe9ODlYAueWGBg+Ow0fnYArqxXA2OgJQAaW0YxzkUADmqDp+MA681gDaforArIEeALruXj7yUHzhUOyhclQQ-MisHmlGZShBqADqwDCiAGKm5pbW6X6OzjCuJRW+-lU2dnZhEfw6UenDfhZWQ8MLHU4ubn0L-QHBeeGRE34FRb0pa+t+mdm5sTBKCaXe6+X3CylPj+1y5w45IVc3ialPNIAX3cDHoMAAnlY4KoHDAwHCZPIAJKsbB8KCw+FwgA8kKsEGwcG44H4wjEEmQAD4wSx2PAIHCETAzFjmUiDEZsM0dJszHEEu4ZFS4Ho4J8cnB8XxCZVAjJpLJCALAoQ4ECnpsanVGjzuWZeUEzAc+MUhdIRSrzSKxRLWMYdNgyKg+EKgfR3cw2KhCnw5KQIIlGdiYO56EA

### CODE REVIEW

**Comment by fabian-hiller** [comment]
I think there was a reason why I choose `InferInput`. I will investigate it to check if it is a bug and report back to you. Thank you!

**Comment by fabian-hiller** [comment]
Hey, I think the current implementation with `InferInput` is correct and your change with `InferOutput` would be a bug. The reason for this is that when you call the function, the argument you pass must match the input of the schema, not the output. The argument schema will then parse the input and pass its value to the function. I know this is hard to follow, but [here](https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6BjCAO1XgGUYphWBzACoQAcgFcQFAKZROTABZSQyOAF5MAOjDAwUgBT04mjrwF6AlABpDmgCZSmoMhetGMGnsnbZoIPeMkZK3pzBmY2DjgAMTFWJhhgNjlFZTVNbV0DNw1sWPjE1hcbd2QoflQ9dxgxMFJ9AG1uUyFRCWlZBSVkAF1zYOyoKWqodkqNVjaZC2DQxhZ2eFsIMQo6gPa09zBS1H0YuISkzuVLOEqyMSkALjgJwKhzNQA+TAupRDgAJlnwhbhB1BiUjwdRLFZrSZQPSET4aACshFCcAA9Mi4AAeAC0mLggkUKDKcBAYki0hQcAABiY+PwKbdqOSKXd2hTfqgIHUNKQIPw9ACgTAkai4HD6EA) is an example that demonstrates this case.

```ts
import * as v from 'valibot';

const StringToNumberSchema = v.pipe(
  v.string(),
  v.decimal(),
  v.transform(Number)
);

const FunctionSchema = v.pipe(
  v.function(),
  v.args(v.tuple([StringToNumberSchema])),
  v.returns(v.number())
);

const doubleNumber = v.parse(FunctionSchema, (value: number) => value * 2);

const result = doubleNumber('2.5'); // <-- The arg must be a `string` not a `number`
console.log(result); // 5
```

**Comment by shulaoda** [comment]
Got it. Thanks again!
